File: claw-os-sdk/go/objects_test.go
package clawossdk

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

type objectVectors struct {
	Valid []struct {
		Reference ObjectRef `json:"reference"`
		URI       string    `json:"uri"`
	} `json:"valid"`
	InvalidURIs []string    `json:"invalid_uris"`
	InvalidRefs []ObjectRef `json:"invalid_refs"`
	Limits      []struct {
		Field string `json:"field"`
		Unit  string `json:"unit"`
		Count int    `json:"count"`
		Bytes int    `json:"bytes"`
	} `json:"limits"`
	WireCases []struct {
		Value json.RawMessage `json:"value"`
		Code  *string         `json:"code"`
		Path  *string         `json:"path"`
	} `json:"wire_cases"`
}

func loadObjectVectors(t *testing.T) objectVectors {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "wire", "v1", "object_ref.vectors.json"))
	if err != nil {
		t.Fatal(err)
	}
	var vectors objectVectors
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatal(err)
	}
	return vectors
}

func objectReference() ObjectRef {
	return ObjectRef{AppId: "notes", ObjectType: "note", ObjectId: "x"}
}

func setObjectField(t *testing.T, value *ObjectRef, field, text string) {
	t.Helper()
	switch field {
	case "app_id":
		value.AppId = text
	case "object_type":
		value.ObjectType = text
	case "object_id":
		value.ObjectId = text
	case "revision":
		value.Revision = &text
	default:
		t.Fatalf("unknown fixture field %s", field)
	}
}

func TestObjectReferenceSharedCanonicalVectors(t *testing.T) {
	for _, entry := range loadObjectVectors(t).Valid {
		uri, err := FormatReference(entry.Reference)
		if err != nil || uri != entry.URI {
			t.Fatalf("FormatReference = %q, %v; want %q", uri, err, entry.URI)
		}
		parsed, err := ParseReference(uri)
		if err != nil || !reflect.DeepEqual(parsed, entry.Reference) {
			t.Fatalf("ParseReference(%q) = %+v, %v; want %+v", uri, parsed, err, entry.Reference)
		}
	}
}

func TestObjectReferenceSharedInvalidVectors(t *testing.T) {
	vectors := loadObjectVectors(t)
	for _, uri := range vectors.InvalidURIs {
		if _, err := ParseReference(uri); err == nil {
			t.Fatalf("accepted invalid URI %q", uri)
		}
	}
	for _, value := range vectors.InvalidRefs {
		if _, err := FormatReference(value); err == nil {
			t.Fatalf("accepted invalid reference %+v", value)
		}
	}
}

func TestObjectReferenceSharedByteLimits(t *testing.T) {
	for _, limit := range loadObjectVectors(t).Limits {
		text := strings.Repeat(limit.Unit, limit.Count)
		if len(text) != limit.Bytes {
			t.Fatalf("%s has %d bytes, want %d", limit.Field, len(text), limit.Bytes)
		}
		value := objectReference()
		setObjectField(t, &value, limit.Field, text)
		uri, err := FormatReference(value)
		if err != nil {
			t.Fatal(err)
		}
		parsed, err := ParseReference(uri)
		if err != nil || !reflect.DeepEqual(parsed, value) {
			t.Fatalf("boundary round trip = %+v, %v", parsed, err)
		}
		setObjectField(t, &value, limit.Field, text+"x")
		if _, err := FormatReference(value); err == nil {
			t.Fatalf("accepted oversized %s", limit.Field)
		}
	}
}

func TestObjectReferenceMaximumComponentsAndURICeiling(t *testing.T) {
	revision := strings.Repeat("\u00e9", 64)
	value := ObjectRef{
		AppId:      strings.Repeat("a", 128),
		ObjectType: strings.Repeat("a", 64),
		ObjectId:   strings.Repeat("\u00e9", 512),
		Revision:   &revision,
	}
	uri, err := FormatReference(value)
	if err != nil || len(uri) != 3669 {
		t.Fatalf("maximum URI: length=%d, error=%v", len(uri), err)
	}
	if _, err := ParseReference(uri); err != nil {
		t.Fatal(err)
	}
	if _, err := ParseReference("app://notes/note?id=" + strings.Repeat("x", 4096)); err == nil {
		t.Fatal("accepted an oversized URI")
	}
}

func TestObjectReferenceInvalidUTF8NeverBecomesAnotherIdentity(t *testing.T) {
	for _, text := range []string{
		string([]byte{0xff}),
		string([]byte{0xed, 0xa0, 0x80}),
		string([]byte{0xf4, 0x90, 0x80, 0x80}),
	} {
		for _, field := range []string{"object_id", "revision"} {
			value := objectReference()
			setObjectField(t, &value, field, text)
			if _, err := FormatReference(value); err == nil {
				t.Fatalf("accepted invalid UTF-8 %s", field)
			}
		}
		if _, err := ParseReference("app://notes/note?id=" + text); err == nil {
			t.Fatal("accepted invalid UTF-8 URI")
		}
	}
}

func TestObjectReferenceGeneratedDecoderSharedContract(t *testing.T) {
	for _, entry := range loadObjectVectors(t).WireCases {
		value := decodeWireValue(t, string(entry.Value))
		err := ValidateObjectRef(value)
		if entry.Code != nil {
			wireErr, ok := err.(*WireDecodeError)
			if !ok || wireErr.Code != *entry.Code || entry.Path == nil || wireErr.Path != *entry.Path {
				t.Fatalf("wire error = %#v, expected code=%v path=%v", err, entry.Code, entry.Path)
			}
			continue
		}
		if err != nil {
			t.Fatal(err)
		}
		var typed ObjectRef
		if err := json.Unmarshal(entry.Value, &typed); err != nil {
			t.Fatal(err)
		}
		encoded, err := json.Marshal(typed)
		if err != nil || !reflect.DeepEqual(decodeWireValue(t, string(encoded)), value) {
			t.Fatalf("wire round trip = %s, %v", encoded, err)
		}
	}
}

func TestObjectReferenceLegacyCompatibilityAndOptionalRevision(t *testing.T) {
	body := `{"id":"notes","version":"1.0.0","name":{"en":"Notes"}}`
	var manifest Manifest
	if err := json.Unmarshal([]byte(body), &manifest); err != nil {
		t.Fatal(err)
	}
	encoded, err := json.Marshal(manifest)
	if err != nil || !reflect.DeepEqual(decodeWireValue(t, string(encoded)), decodeWireValue(t, body)) {
		t.Fatalf("legacy manifest changed: %s, %v", encoded, err)
	}
	value := objectReference()
	if _, err := FormatReference(value); err != nil {
		t.Fatal(err)
	}
	empty := ""
	value.Revision = &empty
	if _, err := FormatReference(value); err == nil {
		t.Fatal("a present empty revision must not become an absent revision")
	}
}

File: claw-os-sdk/node/src/objects.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { resolve } from "node:path";

import { objects } from "./index";
import { type Manifest, type ObjectRef, WireDecodeError, validateObjectRef } from "./generated";

interface Vectors {
  valid: Array<{ reference: ObjectRef; uri: string }>;
  invalid_uris: string[];
  invalid_refs: ObjectRef[];
  limits: Array<{
    field: "app_id" | "object_type" | "object_id" | "revision";
    unit: string;
    count: number;
    bytes: number;
  }>;
  wire_cases: Array<{ value: unknown; code: string | null; path: string | null }>;
}

const vectors: Vectors = JSON.parse(
  readFileSync(resolve(__dirname, "../../wire/v1/object_ref.vectors.json"), "utf8"),
);

function reference(): ObjectRef {
  return { app_id: "notes", object_type: "note", object_id: "x" };
}

test("shared canonical object vectors round trip without identity changes", () => {
  for (const entry of vectors.valid) {
    const value = Object.freeze({ ...entry.reference });
    assert.equal(objects.format_reference(value), entry.uri);
    assert.deepEqual(objects.parse_reference(entry.uri), entry.reference);
    assert.equal(objects.formatReference(value), entry.uri);
    assert.deepEqual(objects.parseReference(entry.uri), entry.reference);
  }
});

test("shared invalid object vectors reject permissive URI normalization", () => {
  for (const uri of vectors.invalid_uris) {
    assert.throws(() => objects.parse_reference(uri), objects.ObjectRefError, uri);
  }
  for (const value of vectors.invalid_refs) {
    assert.throws(() => objects.format_reference(value), objects.ObjectRefError);
  }
});

test("shared byte limits accept the exact boundary and reject one more byte", () => {
  for (const limit of vectors.limits) {
    const value = reference();
    const text = limit.unit.repeat(limit.count);
    assert.equal(Buffer.byteLength(text, "utf8"), limit.bytes);
    value[limit.field] = text;
    assert.deepEqual(objects.parse_reference(objects.format_reference(value)), value);
    value[limit.field] = `${text}x`;
    assert.throws(() => objects.format_reference(value), objects.ObjectRefError, limit.field);
  }
});

test("maximum object components fit the URI ceiling", () => {
  const value: ObjectRef = {
    app_id: "a".repeat(128),
    object_type: "a".repeat(64),
    object_id: "\u00e9".repeat(512),
    revision: "\u00e9".repeat(64),
  };
  const uri = objects.format_reference(value);
  assert.equal(Buffer.byteLength(uri, "utf8"), 3669);
  assert.deepEqual(objects.parse_reference(uri), value);
  assert.throws(
    () => objects.parse_reference(`app://notes/note?id=${"x".repeat(4096)}`),
    objects.ObjectRefError,
  );
});

test("unpaired UTF-16 surrogates are never replacement encoded", () => {
  for (const text of ["\ud800", "\udfff", "x\ud800", "\ud800x"]) {
    for (const field of ["object_id", "revision"] as const) {
      const value = reference();
      value[field] = text;
      assert.throws(() => objects.format_reference(value), objects.ObjectRefError);
    }
    assert.throws(() => objects.parse_reference(`app://notes/note?id=${text}`), objects.ObjectRefError);
  }
});

test("generated ObjectRef decoder follows the shared closed contract", () => {
  for (const entry of vectors.wire_cases) {
    if (entry.code === null) {
      assert.doesNotThrow(() => validateObjectRef(entry.value));
    } else {
      assert.throws(
        () => validateObjectRef(entry.value),
        (error: unknown) =>
          error instanceof WireDecodeError && error.code === entry.code && error.path === entry.path,
      );
      assert.throws(
        () => Reflect.apply(objects.format_reference, undefined, [entry.value]),
        objects.ObjectRefError,
      );
    }
  }
});

test("legacy manifests and unrevisioned references remain compatible", () => {
  const manifest: Manifest = { id: "notes", version: "1.0.0", name: { en: "Notes" } };
  assert.equal(Object.hasOwnProperty.call(manifest, "objects"), false);
  const value = objects.parse_reference("app://notes/note?id=x");
  assert.equal(Object.hasOwnProperty.call(value, "revision"), false);
  assert.equal(Object.hasOwnProperty.call(value, "wire_version"), false);
});
